### ISSUE TITLE
How to encode a chain of certs?

### ISSUE THREAD

**Issue by kwatsen** [netconf-wg/zero-touch#16]
The draft currently states that the owner-certificate is just a single certificate

```
      container owner-certificate {
        leaf certificate {
            "An X.509 v3 certificate structure as specified by RFC
             5280, Section 4 encoded using the ASN.1 distinguished
             encoding rules (DER), as specified in ITU-T X.690.";
        }
        leaf issuer-crl {
          type binary;
          description
            "An CRL structure as specified by RFC 5280, Section 5
             encoded using the ASN.1 distinguished encoding rules
             (DER), as specified in ITU-T X.690.";
        }
      }
```

But the owner certificate actually needs to be presented along with its chain of intermediate certificates leading up to the trust anchor certificate known the to manufacturer's devices.  Here are some options:
1. use an ordered-by-user leaf-list of the X.509v3 structures encoded using DER
2. use a PEM file (string) that can encode multiple certificate 
3. use a PKCS#12 structure from RFC 7292 to encode a bundle of certs?

Thoughts?

Notes: 
- Option #1 is what we current use in system-keychain model, but I think that it’s clunky and should be replaced.
- OpenSSL can convert between #2 and #3 well enough.

Kent

FWIW, here's how the text might look like with a PKCS#12 structure:

```
        leaf certificate {
          type binary;
          mandatory true;
          description
            "A PFU PDU structure as specified by RFC 7292, Section 4 
             encoded using the ASN.1 distinguished encoding rules 
             (DER), as specified in ITU-T X.690.  This is structure
             encodes the owner ceritifcate and any intermediate 
             certificates leading up to the trust anchor certificate
             known to the manufacture's devices.  The owner certificate
             itself encodes a manufacturer-assigned value identifying
             the owner.  This identifier must match the owner 
             identifier encoded in the ownership voucher.";
          reference
            "RFC 7292:
               Personal Information Exchange Syntax v1.1.
             ITU-T X.690:
                Information technology – ASN.1 encoding rules:
                Specification of Basic Encoding Rules (BER),
                Canonical Encoding Rules (CER) and Distinguished
                Encoding Rules (DER).";
        }
```

**Comment by kwatsen** [comment]
From: "Ersue, Mehmet (Nokia - DE/Munich)" <mehmet.ersue@nokia.com>
Date: Saturday, August 13, 2016 at 8:57 AM
To: Kent Watsen <kwatsen@juniper.net>, "netconf@ietf.org" <netconf@ietf.org>
Subject: RE: [Netconf] zerotouch/16: How to encode a chain of certs?

Dear All,
 
technical spoken I’m in favor of supporting all formats with a ‘format’ leaf.
 
@All: Please comment on the proposal from Kent below.
 
Mehmet

**Comment by kwatsen** [comment]
[moving to edit state, following posted to list]

So both Mahesh and Mehmet (as contributors I believe) selected option #6.  No other responses were received.  Thus I will move forward with option #6.
 
Thanks,
Kent

**Comment by kwatsen** [comment]
The PKCS#12 format is giving me trouble.   I dumped its ASN.1 content to ensure I was documenting the structure correctly, only to discover that it was password-encrypted, even though I hit <return> for the password (note: this is because it’s using a salt value).  I spent over an hour looking for a way to remove the password completely, but to no avail.  Unless anyone has an idea, I think we should give up on the pkcs#12 format.
 
So now it’s just PEM and PKCS#7, but between the two, PKCS#7 is by far the better choice.   The idea of supporting more than one format was only really interesting when pkcs#12 was in the mix.  Now I’m not sure if it’s worth it any more.   So, should we stick with:
 
              +--ro certificate
                  +--ro format    enumeration    <-- choices are “pem” and “pkcs7”
                  +--ro data      binary                   <-- encoding depends on ‘format’ field
 
or
 
              +--ro certificate  binary                <-- hardcoded to pkcs#7
 
 
 
Mehmet/Mahesh, since you two selected option #6 (see below) before, I’m especially interested in your opinions.
 
Thanks,
Kent
